### PR TITLE
Feature/ajv transform before validation

### PR DESCRIPTION
**Pull request by elton-okawa**
Solves #471

### CODE REVIEW

File: ajv/src/ajv.ts
Comment (elton-okawa) [inline ajv/src/ajv.ts:66]: Ajv resolver was not strictly following the root docs that states that default mode is `async`

I kept the default value as `sync` in order to not change actual behavior for now

**Review by jorisre — CHANGES_REQUESTED** [review]
Thank for the PR, we really appreciate it. Well tested and documented 🚀 
I would like to avoid introduce a breaking change. Is there any other way to transform data with AJV?
 
Waiting for @bluebill1049 review.

File: ajv/src/ajv.ts
Comment (elton-okawa) [reply ajv/src/ajv.ts:66]: I totally agree with you, but previous resolver logic basically set it as `sync` by default:
```
(schema, schemaOptions, resolverOptions = {}) =>
...
{ $async: resolverOptions && resolverOptions.mode === 'async' }
```

Can we consider it as bug and set default as `async`? Some users who inadvertently relied on it might have problems

**Comment by elton-okawa** [comment]
Hi @jorisre, thanks for the kind words

`ajv` do have ways to transform data:
* [Ajv docs - Modifying data during validation](https://ajv.js.org/guide/modifying-data.html) - built in options to use for removing additional properties, assigning schema default values and coercing types
* [Ajv docs - User defined keywords](https://ajv.js.org/keywords.html) - allows you to define custom keywords to perform some transformation on specific fields
* [ajv-keywords](https://github.com/ajv-validator/ajv-keywords) - extension package from `ajv` maintainers with more keywords

At least for my feature request use case, those possibilities didn't work well mainly because I received a ready to use `ajv schema` to build a form and `keywords` that modify data must live in the schema. One possibility would be dynamically modify given schema but it didn't seem a good idea.

I'd like to understand why adding an optional `transform` option would be a breaking change, I tried to make sure that the behavior stayed the same if you didn't provide it:
```
const valid = validate(transform ? transform(values) : values);
```

File: ajv/src/ajv.ts
Comment (jorisre) [reply ajv/src/ajv.ts:66]: Yes, we can consider it as a bug and set default as `async`

**Comment by jorisre** [comment]
It would be a breaking change because the `transform` option should be is `schemaOptions` instead of `factoryOptions`. So:

```tsx
export type Resolver = <T extends FieldValues>(
  schema: Ajv.JSONSchemaType<T>,
  schemaOptions?: Ajv.Options, // ⬅️ you've to add a transform option here (like the classValidator resolver) but it'll introduce a breaking change
  factoryOptions?: { mode?: 'async' | 'sync' },
) => <TContext>(
  values: T,
  context: TContext | undefined,
  options: ResolverOptions<T>,
) => Promise<ResolverResult<T>>;
```

Ajv looks powerful but there are several ways to use it (with many plugins/modules) so it's complicated to create a resolver that supports all AJV's uses case.

I'm thinking to keep the resolver simple and you can your own.
At the same time, we can discuss, and iterate in order to find a better way to support data transformation that fits all the cases.

**Comment by elton-okawa** [comment]
I see, now I fully understood why it'd be a breaking change, thanks!

I agree with you about keeping the resolver as simpler as possible because it might become a nightmare to maintain in the future.
Let's wait bluebill's opinion about adding the `transform`  option like `class-validator` does

**Comment by bluebill1049** [comment]
Hi guys, sorry for the late reply. Could we hold the `transform` option? If this is a requirement, could we keep this a custom resolver at your app level?

**Comment by elton-okawa** [comment]
Hi @bluebill1049 sure! We can hold the `transform` option changes without any problems.

**Pull request closed without merging**
